### PR TITLE
emscriptenStdenv: create writable cache directory

### LINKED ISSUE #139943: Emscripten tries to write to /nix/store
### Describe the bug
I am trying to compile a project which (among other things) uses emscripten.  
I have created a custom `shell.nix` file to create a development environment for the project.  
Executing the makefile fails when it gets to compiling the emscripten part of the project with the following error message:  
`OSError: [Errno 30] Read-only file system: '/nix/store/ajv94air67dbyjramrjlv0y6j969hsyx-emscripten-2.0.27/share/emscripten/cache/ports'`

### Steps To Reproduce
The project makes use of the open source project imgui. One of it's examples can be used to reproduce the error message.  

1. `git clone https://github.com/ocornut/imgui.git`
2. `cd examples/example_emscripten_opengl3`
3. create a file named `shell.nix` with the following contents:  
   ```shell.nix
   {pkgs ? import <nixpkgs> {}}: pkgs.mkShell {
     nativeBuildInputs = with pkgs; [
       gnumake gcc
     ];

     buildInputs = with pkgs; [
       emscripten
     ];
   }
   ```
4. run `nix-shell`
5. run `make`

### Expected behavior
The command succeeds resulting in a compiled executable of the example project.

### Real behavior
The execution of `make` fails with the following error:  
```
$ make
em++ -DIMGUI_DISABLE_FILE_FUNCTIONS -I../.. -I../../backends -Wall -Wformat -Os -s USE_SDL=2 -s DISABLE_EXCEPTION_CATCHING=1  -c -o main.o main.cpp
Traceback (most recent call last):
  File "/nix/store/ajv94air67dbyjramrjlv0y6j969hsyx-emscripten-2.0.27/share/emscripten/em++.py", line 14, in <module>
    sys.exit(emcc.run(sys.argv))
  File "/nix/store/ajv94air67dbyjramrjlv0y6j969hsyx-emscripten-2.0.27/share/emscripten/emcc.py", line 1074, in run
    linker_inputs = phase_compile_inputs(options, state, newargs, input_files)
  File "/nix/store/dqxic3j7csd4ywn94n4smmnz55p039g3-python3-3.9.6/lib/python3.9/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/nix/store/ajv94air67dbyjramrjlv0y6j969hsyx-emscripten-2.0.27/share/emscripten/emcc.py", line 2436, in phase_compile_inputs
    compile_source_file(i, input_file)
  File "/nix/store/ajv94air67dbyjramrjlv0y6j969hsyx-emscripten-2.0.27/share/emscripten/emcc.py", line 2416, in compile_source_file
    cmd = get_clang_command(input_file)
  File "/nix/store/ajv94air67dbyjramrjlv0y6j969hsyx-emscripten-2.0.27/share/emscripten/emcc.py", line 2357, in get_clang_command
    return get_compiler(use_cxx(src_file)) + get_cflags(options, state.orig_args) + compile_args + [src_file]
  File "/nix/store/ajv94air67dbyjramrjlv0y6j969hsyx-emscripten-2.0.27/share/emscripten/emcc.py", line 879, in get_cflags
    system_libs.add_ports_cflags(cflags, settings)
  File "/nix/store/ajv94air67dbyjramrjlv0y6j969hsyx-emscripten-2.0.27/share/emscripten/tools/system_libs.py", line 1969, in add_ports_cflags
    port.get(Ports, settings, shared)
  File "/nix/store/ajv94air67dbyjramrjlv0y6j969hsyx-emscripten-2.0.27/share/emscripten/tools/ports/sdl2.py", line 23, in get
    ports.fetch_project('sdl2', 'https://github.com/emscripten-ports/SDL2/archive/' + TAG + '.zip', SUBDIR, sha512hash=HASH)
  File "/nix/store/ajv94air67dbyjramrjlv0y6j969hsyx-emscripten-2.0.27/share/emscripten/tools/system_libs.py", line 1765, in fetch_project
    fullname = os.path.join(Ports.get_dir(), name)
  File "/nix/store/ajv94air67dbyjramrjlv0y6j969hsyx-emscripten-2.0.27/share/emscripten/tools/system_libs.py", line 1746, in get_dir
    shared.safe_ensure_dirs(dirname)
  File "/nix/store/ajv94air67dbyjramrjlv0y6j969hsyx-emscripten-2.0.27/share/emscripten/tools/utils.py", line 27, in safe_ensure_dirs
    os.makedirs(dirname, exist_ok=True)
  File "/nix/store/dqxic3j7csd4ywn94n4smmnz55p039g3-python3-3.9.6/lib/python3.9/os.py", line 225, in makedirs
    mkdir(name, mode)
OSError: [Errno 30] Read-only file system: '/nix/store/ajv94air67dbyjramrjlv0y6j969hsyx-emscripten-2.0.27/share/emscripten/cache/ports'
make: *** [Makefile:70: main.o] Error 1
```

During building of the example project make tries to execute the command:  
`em++ -DIMGUI_DISABLE_FILE_FUNCTIONS -I../.. -I../../backends -Wall -Wformat -Os -s USE_SDL=2 -s DISABLE_EXCEPTION_CATCHING=1  -c -o main.o main.cpp`
which fails with the error:  
`OSError: [Errno 30] Read-only file system: '/nix/store/ajv94air67dbyjramrjlv0y6j969hsyx-emscripten-2.0.27/share/emscripten/cache/ports'`

### Notify maintainers
@qknight 
@matthewbauer 

### Metadata
- system: `"x86_64-linux"`
- host os: `Linux 5.13.4, NixOS, 21.11 (Porcupine)`
- multi-user?: `yes`
- sandbox: `yes`
- version: `nix-env (Nix) 2.3.15`
- channels(root): `"nixos-21.11pre318500.51bcdc4cdaa"`
- channels(mbust): `""`
- nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

### PR DESCRIPTION
**Pull request by willcohen**
###### Description of changes

In the `configurePhase` of `emscriptenStdenv`, create a subdirectory that can be used for the internal emscripten cache. Fixes #139943 and #170621.

Issue broke zlib with the update to 3.0 in #149083.

Additionally this modified the derivation for `zlib` to just use the `emscriptenStdenv` configure phase, and then adds the same changes to the other derivations that make larger modifications.

@arximboldi @busti 
@qknight @matthewbauer

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by willcohen** [comment]
Additionally, referencing ZHF: #172160

**Comment by willcohen** [comment]
`nixpkgs-review` is failing to build anything new but using `nix-build $NIXPKGS emscriptenPackages.zlib`, `.libxml2`, `.json_c` shows successful builds for me. `.xmlmirror` has more failures but that appears to involve the software being more out of date and requires a little more work.

**Comment by arximboldi** [comment]
This is great, thank you so much @willcohen !!

**Comment by willcohen** [comment]
@NixOS/release-engineers 
I know that this doesn't technically reduce the red number on hydra since emscripten packages don't seem to build there automatically , but it does still seem in the spirit of ZHF to fix the build of at least 3 of the 4 `emscriptenPackages`, since they've all been broken since the upgrade to emscripten 3.

File: pkgs/development/em-modules/generic/default.nix [hunk 1]
```diff
@@ -1,4 +1,4 @@
-{ pkgs, lib, emscripten, python2 }:
+{ pkgs, lib, emscripten, python }:
```
Comment (Artturin) [inline pkgs/development/em-modules/generic/default.nix:1]: why was this changed?

Comment (willcohen) [reply pkgs/development/em-modules/generic/default.nix:1]: That's a good question @Artturin. `git blame` places the use of python2 with 41faf39, which says that it also hasn't been tested with python3 yet. Given that `emscripten` has required python3 [for a while](https://github.com/emscripten-core/emscripten/issues/7198), I figured it'd be a reasonable time to update this too. Happy to switch it back to make the PR more conservative, but I'm also not sure what role python2 would play anymore in this stdenv.

Comment (Artturin) [reply pkgs/development/em-modules/generic/default.nix:1]: `python` is an alias to `python2` so thats why im asking

if it should be python3 then it should be `python3`

Comment (willcohen) [reply pkgs/development/em-modules/generic/default.nix]: Many apologies, I did not know that! Force-pushed to replace with python3. Build still works for me for the three non-broken packages.

**Comment by willcohen** [comment]
Thank you!

**Comment by pca006132** [comment]
Hi, I wonder if it is possible to document this in `doc/languages-frameworks/emscripten.section.md`, so we will know what to do when we override the configure phase, e.g. for building cmake based projects.
